src/common/transport.cc

```diff
@@ -66,7 +66,7 @@ namespace Pistache
             }
             int fd = peer->fd();
             {
-                Guard guard(toWriteLock);
+                std::scoped_lock lock(toWriteLock);
```

Comment:
`std::scoped_lock` can be used with one or multiple mutexes, and does not impact performance. It removes the question "what lock should I use?", since  `scoped_lock` is full replacement for `lock_guard`. No potential mistakes

Reply:
Yes, by replacement they mean for the situation where you have multiple mutices. 

I thought about this and also asked around. In general always use `std::unique_lock` unless you need something special. A helpful heuristic is to always use the simplest class that you need. The `std::unique_lock` is simple, powerful, and well integrated into conditional variables as well. They are also differentiated for performance reasons. `std::unique_lock` doesn't need to make any branching for dealing with multiple locks or solving deadlocks. So you probably should be using `std::unique_lock`.

Reply:
I think that you mean `lock_guard` (instead of `unique_lock`). 

There's no performance penalty in using `scoped_lock`, as the number of mutexes to lock is determined at compile time. Also, for example, libstdc++ and libc++ have a specialised `lock_guard` implementation that's used when you need to lock a single mutex, and is exactly the same as the implementation of `unique_lock`.

Some static analysis tools even warn you when you use `unique_lock` in C++17 and above, like [SonarQube](https://rules.sonarsource.com/cpp/RSPEC-5997).

`scoped_lock` is as simple as `unique_lock`, while solving the only issue the old class had. I wouldn't be surprised if the latter would get deprecated.

This isn't really a big deal, and if you prefer to use `std::lock_guard` that's ok. I can't find a rationale for it though.

Reply:
No I actually meant `std::unique_lock`. Same with smart pointers. If in doubt, use `std::unique_ptr`.

You can't say for certain that there is no performance penalty in using it because we're not in control of every implementation. The standard library that ships with GCC is just one of many.

What we do know is that the _intention_ behind `std::scoped_lock` was to address the problem of deadlocks with multiple mutices. When the user uses it, what they are saying in effect to the library designers is this: "You may need to do something special to deal with the problem of deadlocks. I'm willing to pay a small price for it because I can't do that with other locks." Whether the implementation actually carries with it no additional performance penalty, even with only a single lock, is beyond the user's control (and sometimes knowledge as well).

If `std::scoped_lock` was actually a drop in replacement for `std::unique_lock` or `std::lock_guard` when using a single mutex, the standards committee would mark the latter two as deprecated and just tell everyone to use the former.

Reply:
Ok then, I'll restore `std::lock_guard`. 

Why `unique_lock`? It is way more complex than `lock_guard`

Reply:
Because it's the simplest of all of them. Always use the simplest class you need, but not any simpler.

Reply:
I think that all this locks and guards are confusing us. [`std::lock_guard`](https://en.cppreference.com/w/cpp/thread/lock_guard) is the simple one, while [`std::unique_lock`](https://en.cppreference.com/w/cpp/thread/unique_lock) is the one that can be unlocked manually etc.

Reply:
Then use `std::lock_guard`. I personally prefer that one, though it is less general purpose than `std::unique_lock`.